app/web_framework/cluster_application.py

```diff
@@ -15,6 +52,33 @@ def get_all_handlers(root_route, default_params):
         :rtype: list [tuple (str, tornado.web.RequestHandler, dict)]
         """
         all_route_nodes = [root_route] + root_route.descendants()
+
+        # Map regex to all the routes that have the same regex pattern.
+        route_map = {}
+        for route in all_route_nodes:
+            route_path = route.regex()
+            if route_path not in route_map:
+                route_map[route_path] = []
```

Comment:
What kind of issues? The regex mappings are used to resolve any ambiguity for any routes that has the same path, so if the regex for v1 and v2 are different (`'artifacts.zip'` and `'artifacts.zip$'`) then the routes should just be able to handle that normally, right?